File: Week3/297.java
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if(root!=null){
            String res="";
            Queue<TreeNode> q=new LinkedList<>();
            if (root!=null) q.offer(root);
            while(!q.isEmpty()) {
                TreeNode node=q.poll();
                if (node!=null){
                    q.offer(node.left);
                    q.offer(node.right);
                    res+=Integer.toString(node.val)+",";
                }else{
                    res+="null,";
                }
            }
            res=res.substring(0,res.length()-1);
            return "["+res+"]";
        }
        return "";
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if(data.length()>2){
            String[] lst=data.substring(1, data.length()-1).split(",");
                TreeNode root=new TreeNode(Integer.valueOf(lst[0]));
                Integer cnt=1;
                Queue<TreeNode> q=new LinkedList<>();
                q.offer(root);
                while(cnt<lst.length){
                    TreeNode node=q.poll();
                    if (!lst[cnt].equals("null")){
                        node.left=new TreeNode(Integer.valueOf(lst[cnt]));
                        q.offer(node.left);
                    }
                    cnt+=1;
                    if (cnt<lst.length && !lst[cnt].equals("null")){
                        node.right=new TreeNode(Integer.valueOf(lst[cnt]));
                        q.offer(node.right);
                    }
                    cnt+=1;
                }
                return root;
        }
        return null;
    }
}

File: Week3/297.py
class Codec:
    
    def serialize(self, root):
        if root:
            res=[]
            queue=[root]
            while queue:
                node=queue[0]
                queue=queue[1:]
                if node:
                    res.append(str(node.val))
                    queue.append(node.left)
                    queue.append(node.right)
                else:
                    res.append("None")
            return "["+",".join(res)+"]"
        return ""
    def deserialize(self, data):
        if data:
            lst=data[1:-1].split(",")
            root=TreeNode(int(lst[0]))
            queue=[root]
            cnt=1
            while cnt<len(lst):
                node=queue[0]
                queue=queue[1:]
                if lst[cnt]!="None":
                    node.left=TreeNode(int(lst[cnt]))
                    queue.append(node.left)
                cnt+=1
                if lst[cnt]!="None":
                    node.right=TreeNode(int(lst[cnt]))
                    queue.append(node.right)
                cnt+=1
            return root
        return